<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Tutorial</title>

<script src="tutorial_files/header-attrs-2.26/header-attrs.js"></script>
<script src="tutorial_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="tutorial_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="tutorial_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="tutorial_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="tutorial_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="tutorial_files/navigation-1.1/tabsets.js"></script>
<link href="tutorial_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="tutorial_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Tutorial</h1>

</div>


<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Welcome to <code>RiboCrypt</code> <code>RiboCrypt</code> is an R
package for interactive visualization in genomics.
<code>RiboCrypt</code> works with any NGS-based method, but much
emphasis is put on Ribo-seq data visualization.</p>
<p>This tutorial will walk you through usage of the app.</p>
<p><code>RibCrypt</code> app currently supports creating interactive
browser views for NGS tracks, using ORFik, Ribocrypt and massiveNGSpipe
as backend.</p>
</div>
<div id="browser" class="section level1">
<h1>Browser</h1>
<p>The browser is the main coverage plot display page. It contains a
click panel on the left side and display panels on the right. It
displays coverage of NGS data in either transcript coordinates
(default), or genomic coordinates (like IGV). Each part will now be
explained:</p>
<p><img src="../images/tutorial_browser_browser.png" width="1000px" height="500px" style="display: block; margin: auto auto auto 0;" /></p>
<div id="display-panel-browser" class="section level2">
<h2>Display panel (browser)</h2>
<p>The display panel shows the primary settings, (study, gene, sample,
etc), the possible select boxes are:</p>
<div id="exp" class="section level3">
<h3>Experiment selection</h3>
<ul>
<li>Select an organism: Either select “ALL” to keep all experiments, or
select a specific organism to select display only that subset of
experiments in experiment select tab.</li>
<li>Select an experiment: The experiments contain study names combined
with organism (some studies are multi species, so sometimes one study
have multiple experiments). Select which one you want. There also exist
merged experiments (all samples merged for the organism, etc)</li>
</ul>
</div>
<div id="gene" class="section level3">
<h3>Gene selection</h3>
<ul>
<li>Select a gene: A gene can be selected currently using:
<ul>
<li>Gene id (ENSEMBL)</li>
<li>Gene symbol (hgnc, etc)</li>
</ul></li>
<li>Select a transcript: A transcript isoform of the given gene above,
default is Ensembl canonical isoform. Can be selected using:
<ul>
<li>Transcript id (ENSEMBL)</li>
</ul></li>
</ul>
</div>
<div id="lib" class="section level3">
<h3>Library selection</h3>
<p>Each experiment usually have multiple libraries. Select which one to
display, by default if you select multiple libraries they will be shown
under each other.</p>
<p>Library are by default named:</p>
<ul>
<li>Library type (RFP, RNA etc),</li>
<li>Condition (WT, KO (wild type, knock out ) etc)</li>
<li>Stage/timepoint (5h, 1d (5 hours, 1 day) etc)</li>
<li>fraction (chx, cytosolic, ATF4 (ribosomal inhibitor, cell fraction,
gene) etc)</li>
<li>replicate (technical/biological replicate number (r1, r2, r3))</li>
</ul>
<p>The resuting name above could be:</p>
<ul>
<li>RFP_WT_5h_chx_cytosolic_r1</li>
</ul>
<p>A normal thing to see is that if condition is KO (knockout), the
fraction column usually contains a gene name (the name of the gene that
was knocked out) Currently, best way to find SRR run number for
respective sample is to go to metadata tab and search for the study.</p>
</div>
<div id="linetype" class="section level3">
<h3>View mode</h3>
<ul>
<li>Select frames display type:
<ul>
<li>lines (single line, most clear for middle distance (&gt; 100
nt))</li>
<li>columns (single point bars, most clear for single nt
resolution)</li>
<li>stacks (Area under curve, stacked, most clear for long distance
(&gt; 1000 nt))</li>
<li>area (Area under curve, with alpha (see-through), most clear for
long distance (&gt; 1000 nt))</li>
</ul></li>
<li>K-mer length: When looking at a large region (&gt; 100nt), pure
coverage can usually be hard to inspect. Using K-mer length &gt; 1 (9 is
a good starting point to try), you can easily look at patterns over
larger regions.</li>
</ul>
</div>
</div>
<div id="display-panel-settings" class="section level2">
<h2>Display panel (settings)</h2>
<p>Here additional options are shown:</p>
<ul>
<li>5’ extension (extend viewed window upstream, outside defined
region)</li>
<li>3’ extension (extend viewed window downstream, outside defined
region)</li>
<li>Custom sequences highlight (Motif search, given in purple color,
support IUPAC, examples: CTG or NTG)</li>
<li>Genomic region (Browse genomic window instead of gene, syntax:
chromosome:start-stop:strand, human/mouse/zebrafish: 1:10000-20000:+ ,
yeast: I:10000-20000:+. Both 1 and chr1 works, conversion will be done
internally)</li>
<li>Genomic View (Activate/deactivate genomic view, giving splice
information and correct positions in genome, but a lot harder to
understand)</li>
<li>Protein structures (If you click the annotation name of a transcript
in the plot panel it will display the alpha-fold protein colored by the
ribo-seq data displayed in the plot panel)</li>
<li>Full annotation (display full annotation or just the tx you
selected)</li>
<li>Summary top track (Add an additional plot track on top, summarizing
all selecte libs)</li>
<li>Select Summary display type (same as frames display type above, but
for the summary track)</li>
<li>Export format (When you hover the plot top right image button, and
click export (the camera button), which format to export as)</li>
</ul>
</div>
<div id="plot-panel" class="section level2">
<h2>Plot panel</h2>
<p>From the options specified in the display panel, when you press
“plot” the data will be displayed. It contains the specific parts:</p>
<ol style="list-style-type: decimal">
<li>Ribo-seq data (top), the single or multi-track data is displayed on
top. By default Ribo-seq is displayed in 3 colors, where</li>
</ol>
<ul>
<li>red is 0 frame, the start frame of reference transcript.</li>
<li>green is +1 frame</li>
<li>blue is +2 frame</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>Sequence track (top middle), displayes DNA sequence when zoomed in
(&lt; 100nt)</li>
<li>Annotation track (middle), the annotation track displays the
transcript annotation, together with black bars that is displayed on top
of the data track.</li>
<li>Frame track (bottom), the 3 frames displayed with given color
bars:</li>
</ol>
<ul>
<li>white (Start codons)</li>
<li>black (Stop codons)</li>
<li>purple (Custom motifs) When zoomed in, the amino acid sequence is
displaced within each frame</li>
</ul>
</div>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>Here we collect the analysis possibilities, which are usually on
whole genome scale.</p>
<div id="codon-analysis" class="section level2">
<h2>Codon analysis</h2>
<p>This tab displays a heatmap of codons dwell times over all genes
selected, for both A and P sites. When pressing “Differential” you swap
to a between library differential codon dwell time comparison (minimum 2
libraries selected is required for this method!)</p>
<div id="display-panel-codon" class="section level3">
<h3>Display panel (codon)</h3>
<p>Study and gene select works same as for browser specified above. In
addition to have the option to specify all genes (default). - Select
libraries (multiple allowed)</p>
<div id="codonfilt" class="section level4">
<h4>Filters</h4>
<ul>
<li>Codon filter value (Minimum reads in ORF to be included)</li>
<li>Codon score, all scores are normalized for both codon and count per
gene level (except for sum):
<ul>
<li>percentage (percentage use relative to max codon, transcript
normalized percentages)</li>
<li>dispersion(NB) (negative binomial dispersion values)</li>
<li>alpha(DMN) (Dirichlet-multinomial distribution alpha parameter)</li>
<li>sum (raw sum, (a very biased estimator, since some codons are used
much more than others!))</li>
</ul></li>
</ul>
</div>
</div>
</div>
<div id="heatmap" class="section level2">
<h2>Heatmap</h2>
<p>This tab displays a heatmap of coverage per readlength at a specific
region (like start site of coding sequences) over all genes
selected.</p>
<p><img src="../images/tutorial_heatmap_browser.png" width="1000px" height="500px" style="display: block; margin: auto auto auto 0;" /></p>
<div id="display-panel-heatmap" class="section level3">
<h3>Display panel (heatmap)</h3>
<p>Study and gene select works same as for browser specified above. In
addition to have the option to specify all genes (default).</p>
<ul>
<li>Select libraries Only 1 library can be selected currently in heatmap
mode.</li>
<li>View region Select one of:
<ul>
<li>Start codon</li>
<li>Stop codon</li>
</ul></li>
<li>Normalization Normalization mode for data display, select one of:
<ul>
<li>transcriptNormalized (each gene counts sum to 1)</li>
<li>zscore (zscore normalization, will give better overview if 1
readlength is extreme)</li>
<li>sum (raw sum of counts, is very sensitive to extreme peaks)</li>
<li>log10sum (log10 sum of counts, is less sensitive to extreme
peaks)</li>
</ul></li>
<li>Min Readlength The minimum readlength to display from library</li>
<li>Max readlength The maximum readlength to display from library</li>
</ul>
</div>
<div id="display-panel-settings-1" class="section level3">
<h3>Display panel (settings)</h3>
<p>Here additional options are shown:</p>
<ul>
<li><p>5’ extension (extend viewed window upstream from point, default
30)</p></li>
<li><p>3’ extension (extend viewed window downstreamfrom point, default
30)</p></li>
<li><p>Extension works like this, first extend to transcript
coordinates.</p></li>
<li><p>After gene end extend in genomic coordinates</p></li>
<li><p>If chromosome boundary is reached, remove those genes from the
full set.</p></li>
</ul>
</div>
</div>
<div id="diffexp" class="section level2">
<h2>Differential gene expression</h2>
<p>Given an experiment with a least 1 design column with two values,
like wild-type (WT) vs knock out (of a specific gene), you can run
differential expression of genes. The output is an interactive plot,
where you can also search for you target genes, making it more useable
than normal expression plots, which often are very hard to read.</p>
<div id="display-panel-differential-expression" class="section level3">
<h3>Display panel (Differential expression)</h3>
<p>Organism and experiment explained above - Differential method: FPKM
ratio is a pure FPKM ratio calculation without factor normalization
(like batch effects), fast and crude check. DESeq2 argument gives a
robust version, but only works for experiments with valid experimental
design (i.e. design matrix must be full ranked, see deseq2 tutorial for
details!) - Select two conditions (which 2 factors to group by)</p>
</div>
<div id="display-panel-settings-2" class="section level3">
<h3>Display panel (settings)</h3>
<ul>
<li>draw unregulated (show dots for unregulated genes, makes it much
slower!)</li>
<li>Full annotation (all transcript isoforms, default is primary isoform
only!)</li>
<li>P-value (sliding bar for p-value cutoff, default 0.05)</li>
<li>export format for plot (explained above)</li>
</ul>
</div>
</div>
<div id="mbrowser" class="section level2">
<h2>Meta Browser</h2>
<p>Display all samples for a specific organism over selected gene. This
tab does not use bigwig files to load (as that would be very slow). It
uses precomputed fst files of coverage over all libraries. Note: Not all
isoforms are computed, by default the longest isoform is computed.</p>
<div id="display-panel-meta-browser" class="section level3">
<h3>Display panel (Meta browser)</h3>
<p>Organism, experiment and gene explained above - Group on: the
metadata column to order plot by - K-means clusters: How many k-means
clusters to use, if &gt; 1, Group will be sorted within the clusters,
but K-means have priority.</p>
</div>
<div id="statistics-tab" class="section level3">
<h3>Statistics tab</h3>
<p>This tab gives the statistics of over representation analysis per
cluster. Using chi squared test, it gives the residuals per term from
metadata (like tissue, cell-line etc). If a value is bigger than +/- 3,
it means it is quite certain this is over represented.</p>
<p>If no clustering was applied, this tab gives the number of items per
metadata term (40 brain samples, 30 kidney samples etc).</p>
</div>
<div id="display-panel-settings-3" class="section level3">
<h3>Display panel (settings)</h3>
<ul>
<li>Normalization (all scores are tpm normalized and log scaled)
<ul>
<li>transcriptNormalized (each sample counts sum to 1) (default)</li>
<li>Max normalized (each position count divided by maximum)</li>
<li>zscore (zscore normalization, (count * mean / sd) variance scaled
normalization)</li>
<li>tpm (raw tpm of counts, is very sensitive to extreme peaks)</li>
</ul></li>
<li>Color theme: Which color theme to use</li>
<li>Color scale multiplier: how much to amplify the color signal (if all
is single color, try to reduce or increase this depending on which color
is the majority)</li>
<li>Sort by other gene: Sort heatmap on expression of another gene
(increasing). A line plot will be shown left that display the expression
of that other gene per row. Also the enrichment analysis is done on bins
of the other genes expression instead of metadata term.</li>
<li>Summary top track (same as in browser, the aggregate of all rows,
useful to see the frame information, which is not represented in the
heatmap)</li>
<li>Split by frame: (Not active yet!): When completed will display the
heatmap with frame information, so rgb colors from white to dark red
etc.</li>
<li>Summary display type: Same as in browser</li>
<li>Region to view: By default uses entire transcript. Can subset to
only view cds or leader+cds etc.</li>
<li>Select a transcript: Which transcript isoforms not all isoforms have
computed fsts. By default the longest isoforms have been computed for
all genes.</li>
<li>K-mer length: Smoothens out the signal by applying a mean sliding
window, default 1 (off)</li>
</ul>
</div>
<div id="requirements" class="section level3">
<h3>Requirements</h3>
<p>This mode is very intensive on CPU, so it requires certain
pre-computed results for the back end. That is namely: - Premade
collection experiments (an ORFik experiment of all experiments per
organism) - Premade collection count table and library sizes (for
normalizations purpose) - Premade fst serialized coverage calculation
per gene (for instant loading of coverage over thousands of
libraries)</p>
<p>Note that on the live app, the human collection (4000 Ribo-seq
samples) takes around 30 seconds to plot for a ~ 2K nucleotides gene,
~99% of the time is spent on rendering the plot, not actual computation.
Future investigation into optimization will be done.</p>
</div>
</div>
<div id="read-length-qc" class="section level2">
<h2>Read length (QC)</h2>
<p>This tab displays a QC of pshifted coverage per readlength (like
start site of coding sequences) over all genes selected.</p>
<div id="display-panel-read-length-qc" class="section level3">
<h3>Display panel (Read length QC)</h3>
<p>The display panel shows what can be specified to display, the
possible select boxes are same as for heatmap above:</p>
</div>
<div id="plot-panel-1" class="section level3">
<h3>Plot panel</h3>
<p>From the options specified in the display panel, when you press
“plot” the data will be displayed. It contains the specific parts:</p>
<p>Top plot: Read length relative usage</p>
<ol style="list-style-type: decimal">
<li>Y-axis: Score</li>
<li>Color: Per frame (red, green, blue)</li>
<li>Facet box: the read length</li>
</ol>
<p>Bottom plot: Fourier transform (3nt periodicity quality, clean peak
means good periodicity)</p>
</div>
</div>
<div id="fastq-qc" class="section level2">
<h2>Fastq (QC)</h2>
<p>This tab displays the fastq QC output from fastp, as a html page.</p>
<div id="display-panel-read-length-qc-1" class="section level3">
<h3>Display panel (Read length QC)</h3>
<p>The display panel shows what can be specified to display, you can
select from organism, study and library.</p>
</div>
<div id="plot-panel-2" class="section level3">
<h3>Plot panel</h3>
<p>Displays the html page.</p>
</div>
</div>
</div>
<div id="metadata" class="section level1">
<h1>Metadata</h1>
<p>Metadata tab displays information about studies and custom
predictions. ## SRA search</p>
<p>Search SRA for full information of supported study</p>
<div id="study-accession-number" class="section level3">
<h3>Study accession number</h3>
<p>Here you input a study accession number in the form of either:</p>
<ul>
<li>SRP</li>
<li>GEO (GSE)</li>
<li>PRJNA (PRJ….)</li>
<li>PRJID (Only numbers)</li>
</ul>
</div>
<div id="output" class="section level3">
<h3>Output</h3>
<p>On top the abstract of the study is displayed, and on bottom a table
of all metadata found from the study is displayed.</p>
</div>
<div id="studies" class="section level2">
<h2>Studies</h2>
<p>Full table of supported studies with information about sample
counts</p>
</div>
<div id="predicted-translons" class="section level2">
<h2>Predicted translons</h2>
<p>Full list of predicted translons on all_merged tracks per
species.</p>
<div id="settings" class="section level3">
<h3>Settings:</h3>
<p>Translon annotation scheme:</p>
<ul>
<li>Prediction algorithm: ORFik::detect_ribo_orfs</li>
<li>translon_types &lt;- c(“uORF”, “uoORF”,“doORF”,“dORF”)</li>
<li>start_codons &lt;- “ATG|TTG|CTG”</li>
<li>additional arguments: add sequence and amino acid sequence per
translon</li>
<li>All other arguments default.</li>
</ul>
</div>
</div>
</div>
<div id="additional-information" class="section level1">
<h1>Additional information</h1>
<p>All files are packed into ORFik experiments for easy access through
the ORFik backend package:</p>
<p>File formats used internally in experiments are:</p>
<ul>
<li>Annotation (gtf + TxDb for random access)</li>
<li>Fasta genome (.fasta, + index for random access)</li>
<li>Sequencing libraries (all duplicated reads are collapsed)
<ul>
<li>random access (only for collapsed read lengths): bigwig</li>
<li>Full genome coverage (only for collapsed read lengths): covRLE</li>
<li>Full genome coverage (split by read lengths): covRLElist</li>
</ul></li>
<li>count Tables (Summarized experiments, r data serialized .rds)</li>
<li>Library size list (Integer vector, .rds)</li>
<li>Precomputed gene coverages per organism: fst (used for
metabrowser)</li>
</ul>
<div id="massivengspipe" class="section level2">
<h2>massiveNGSpipe</h2>
<p>For our webpage the processing pipeline used is massiveNGSpipe which
wraps over multiple tools:</p>
<ol style="list-style-type: decimal">
<li>Fastq files are download with ORFik download.sra</li>
<li>Adapter is detected with either fastqc (sequence detection) and
falls back to fastp auto detection.</li>
<li>Reads are then trimmed with fastp (using the wrapper in ORFik)</li>
</ol>
<ul>
<li>Adapter removal specified</li>
<li>minimum read size (20nt)</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>Read are collapsed (get the set of unique reads and put duplication
count in read header)</li>
<li>Reads are aligned with the STAR aligner (using the wrapper in
ORFik), that supports contamination removal. Settings:</li>
</ol>
<ul>
<li>genomic coordinates (to allow both genomic and transcriptomic
coordinates)</li>
<li>local alignment (to remove unknown flank effects)</li>
<li>minimum read size (20nt)</li>
</ul>
<ol start="6" style="list-style-type: decimal">
<li>When all samples of study are aligned, an ORFik experiment is
created that connects each sample to metadata (condition, inhibitor,
fraction, replicate etc)</li>
<li>Bam files are then converted to ORFik ofst format</li>
<li>These ofst files are then pshifted</li>
<li>Faster formats are then created (bigwig, fst and covRLE) for faster
visualization</li>
</ol>
</div>
<div id="introduction-to-ribo-seq" class="section level2">
<h2>Introduction to Ribo-seq</h2>
<p>If you’re not familiar with terms like “p-shifting” or “p-site
offset”, it’s best to walk through ORFikOverview vignette, especially
chapter 6 “RiboSeq footprints automatic shift detection and
shifting”</p>
<p><a
href="https://bioconductor.org/packages/release/bioc/vignettes/ORFik/inst/doc/ORFikOverview.html#riboseq-footprints-automatic-shift-detection-and-shifting"
class="uri">https://bioconductor.org/packages/release/bioc/vignettes/ORFik/inst/doc/ORFikOverview.html#riboseq-footprints-automatic-shift-detection-and-shifting</a></p>
</div>
<div id="api-for-url-access-and-sharing" class="section level2">
<h2>API for URL access and sharing</h2>
<p>RiboCrypt uses the shiny router API system for creating runable links
and backspacing etc. The API specificiation is the following:</p>
<div id="primary-url" class="section level3">
<h3>Primary url:</h3>
<p><a href="https://ribocrypt.org/"
class="uri">https://ribocrypt.org/</a> (This leads to browser page)</p>
</div>
<div id="page-selection-api" class="section level3">
<h3>Page selection API:</h3>
<p>Page selection is done with “#” followed by the page short name, the
list is the following:</p>
<ul>
<li>broser page (/ or /#browser)</li>
<li>heatmap (/#heatmap)</li>
<li>codon (/#codon)</li>
<li>Differential expression (/#Differential expression)</li>
<li>Periodicity plot (/#periodicity)</li>
<li>fastq QC report (/#fastq)</li>
<li>MetaBrowser (/#MetaBrowser)</li>
<li>SRA search (/#SRA search)</li>
<li>Studies supported (/#Studies)</li>
<li>This tutorial (/#tutorial)</li>
</ul>
<p>Example: <a href="https://ribocrypt.org/#tutorial"
class="uri">https://ribocrypt.org/#tutorial</a> sends you to this
tutorial page</p>
</div>
<div id="parameter-api" class="section level3">
<h3>Parameter API:</h3>
<p>Settings can be specified by using the standard web parameter
API:</p>
<ul>
<li>“?”, Starts the parameter specification</li>
<li>“&amp;”, to combine terms:</li>
</ul>
<p>Example: <a
href="https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272#browser"
class="uri">https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272#browser</a>
will lead you to browser and insert gene ATF4 (all other settings being
default).</p>
<p>A more complicated call would be: <a
href="https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272&amp;tx=ENST00000404241&amp;frames_type=area&amp;kmer=9&amp;go=TRUE&amp;extendLeaders=100&amp;extendTrailers=100&amp;viewMode=TRUE&amp;other_tx=TRUE#browser"
class="uri">https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272&amp;tx=ENST00000404241&amp;frames_type=area&amp;kmer=9&amp;go=TRUE&amp;extendLeaders=100&amp;extendTrailers=100&amp;viewMode=TRUE&amp;other_tx=TRUE#browser</a></p>
<div id="browser-1" class="section level4">
<h4>browser:</h4>
<ul>
<li>dff=all_merged-Homo_sapiens (The Experiment to select: for webpage
it is “study id”_“Organism”)</li>
<li>gene=ATF4-ENSG00000128272 (For webpage it is: “Gene symbol”-“Ensembl
gene id”)</li>
<li>tx=ENST00000404241(isoform)</li>
<li>frames_type=area (plot type)</li>
<li>kmer=9 (window smoothing)</li>
<li>go=TRUE (plot on entry, you do not need to click plot for it to
happen)</li>
<li>extendLeaders=100 (extend 100 nt upstream of 5’ UTR / Leader)</li>
<li>extendTrailers=100 (extend 100 nt downstream of 3’ UTR /
Trailer)</li>
<li>viewMode=TRUE (Genomic coordinates, i.e. with introns)</li>
<li>other_tx=TRUE (Full annotation, show transcript graph for all
genes/isoforms in area)</li>
<li>#browser (open browser window)</li>
</ul>
</div>
</div>
</div>
<div id="about" class="section level2">
<h2>About</h2>
<p>This app is created as a collaboration with:</p>
<ul>
<li>University of Warsaw, Poland</li>
<li>University of Bergen, Norway</li>
</ul>
<p>Main authors and contact:</p>
<ul>
<li>Michal Swirski (Warsaw), email: <a
href="mailto:michal.swirski@uw.edu.pl"
class="email">michal.swirski@uw.edu.pl</a></li>
<li>Håkon Tjeldnes (Bergen), email: <a
href="mailto:hauken_heyken@hotmail.com"
class="email">hauken_heyken@hotmail.com</a></li>
</ul>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
